geo/src/algorithm/outlier_detection.rs

```diff
@@ -0,0 +1,417 @@
+        Self { tree, points }
+    }
+
+    /// The LOF algorithm. `k_neighbours` specifies the number of neighbours to use for local outlier
+    /// classification. The paper linked above (see p. 100) suggests a `k_neighbours` value of 10 - 20
+    /// as a lower bound for "real-world"
+    /// data. The accuracy of the algorithm can be increased (see p. 101) by performing multiple passes
+    /// over the data using a range of `k_neighbours` values, then aggregating these, taking e.g.
+    /// the minimum or the mean.
+    ///
+    /// If `k_neighbours` >= points in the set, input points will be returned with a `score` of 1.
+    /// If there are at least `k_neighbours` duplicate points of an input point, LOF scores can be ∞ or `NaN`.
+    /// It is thus advisable to **deduplicate** or otherwise ensure the uniqueness of the input points.
+    ///
+    /// # Examples
+    ///
+    /// `MultiPoint`:
+    ///
+    /// ```
+    /// use approx::assert_relative_eq;
+    /// use geo::OutlierDetection;
+    /// use geo::{point, MultiPoint};
+    ///
+    /// let v = vec![
+    ///     point!(x: 0.0, y: 0.0),
+    ///     point!(x: 0.0, y: 1.0),
+    ///     point!(x: 3.0, y: 0.0),
+    ///     point!(x: 1.0, y: 1.0),
+    /// ];
+    /// let mp = MultiPoint(v);
+    ///
+    /// let lofscores = mp.outliers(3);
+    ///
+    /// assert_relative_eq!(lofscores[2].score, 3.3333333333333335);
+    /// ```
+    pub fn outliers(&self, kneighbours: usize) -> Vec<LofScore<T>> {
+        lof(self.points, &self.tree, kneighbours)
+    }
+}
+
+fn lof<T>(
+    points: &[Point<T>],
+    tree: &RTree<GeomWithData<Point<T>, usize>>,
+    kneighbours: usize,
+) -> Vec<LofScore<T>>
+where
+    T: GeoFloat + Sum,
+{
+    if points.len() <= kneighbours {
+        // no point in trying to run the algorithm in this case
+        return points
+            .iter()
+            .enumerate()
+            .map(|(idx, _)| LofScore::new(idx, T::one()))
+            .collect();
+    }
+    let knn_dists = points
+        .iter()
+        .map(|point| {
+            tree.nearest_neighbor_iter_with_distance_2(point)
+                .take(kneighbours)
+                .collect()
+        })
+        .collect::<Vec<Vec<_>>>();
+    // calculate LRD (local reachability density) of each point
+    // LRD is the estimated distance at which a point can be found by its neighbours:
+    // count(neighbour_set) / sum(max(point.kTh_dist, point.dist2(other point)) for all points in neighbour_set)
+    // we call this sum-of–max-distances reachDistance
+    let local_reachability_densities: Vec<T> = knn_dists
+        .iter()
+        .map(|neighbours| {
+            // for each point's neighbour set, calculate kth distance
+            let kth_dist = neighbours
+                .iter()
+                .map(|(_, distance)| distance)
+                .last()
+                .unwrap();
+            T::from(neighbours.len()).unwrap()
+                / neighbours
```

Comment:
It seems like we'd get a div-by-0 here when there are sufficient repeated points. Does the algorithm provide sensible results in that case?

Reply:
It returns either infinity or `NaN` at the moment (which is documented). I'm not sure how we can avoid that without affecting performance, though…

Reply:
I've made this possible result extra-clear in the docs, and because this input can't produce a panic I'm OK with it as is.